### ISSUE TITLE
Should u-* parsing special case img srcset?

### ISSUE THREAD

**Issue by tantek** [microformats/microformats2-parsing#7]
Responsive images add the 'srcset' attribute to the 'img' tag which can then provide a variety of image resolutions, both in density and pixel size.

Should u-* on an img tag pay attention to the srcset attribute, and if so, how?

Summary of srcset here https://indieweb.org/srcset

**Comment by snarfed** [comment]
yes! use case: https://github.com/snarfed/bridgy/issues/592#issuecomment-331210211 . a few people want [bridgy publish](https://brid.gy/about#publishing) to support srcset, specifically to publish the largest available image that's under a silo's size limit. e.g. twitter's is 5MB.

cc @jonnybarnes for an example post with `srcset`.

**Comment by jonnybarnes** [comment]
Here’s an initial post that uses `srcset`: https://jonnybarnes.uk/notes/GE

**Comment by jonnybarnes** [comment]
It appears php-mf2 leaves the `srcset` attribute blank: https://pin13.net/mf2/?url=https%3A%2F%2Fjonnybarnes.uk%2Fnotes%2FGE

**Comment by sknebel** [comment]
#2 proposes already to expand images into an object with `url` and `alt` attributes. This object could be expanded to cover srcset as well.

Looking at the [examples linked](https://indieweb.org/srcset#simple_image_pixel_size_approach, from my understanding the `sizes` attribute doesn't need to be preserved since it depends on the design of the specific page and does not transfer over into different displays (a downstream consumer displaying it could insert a `sizes` attribute in their own page, matching their own stylesheets.

based on example HTML from the wiki:
```html
<img 
  class="u-photo"
  src="http://example.com/img-360x540.jpg" 
  alt="img" 
  srcset="img-240x360.jpg 360w, img-360x540.jpg 540w, img-653x980.jpg 980w, img-853x1280.jpg 1280w" 
  sizes="(min-width: 960px) 50vw, 100vw"
/>
``` 
The parser could just pass on the content of `srcset`, leaving the downstream consumer to handle it completely, giving output along the lines of
```json
…
"photo": {
   "value": "http://example.com/img-360x540.jpg",
   "alt": "img",
   "srcset": "img-240x360.jpg 360w, img-360x540.jpg 540w, img-653x980.jpg 980w, img-853x1280.jpg 1280w" 
}
```

, or transform it into some kind of more detailed structure (suggestions at what that'd ideally look like for a consumer? Seems like @snarfed would only want something where the full list of images is easy to get from, since the parameters don't actually matter for bridgy?)

**Comment by sknebel** [comment]
@grantcodes I believe you were interested in this too, any feedback?

**Comment by grantcodes** [comment]
@sknebel yeah, I didn't comment as my use case is very different, and probably not a microformats issue.

I am trying to generate html from mf2 json and very rarely want to show a full resolution image. Your proposal may work for that but honestly I have not thought about it enough to be sure.

**Comment by snarfed** [comment]
friendly nudge! this came up in chat just now: https://chat.indieweb.org/dev/2020-04-03#t1585936229416600

**Comment by dshanske** [comment]
Would suggest sourceset might be best parsed out.

`"srcset": "img-240x360.jpg 360w, img-360x540.jpg 540w, img-653x980.jpg 980w, img-853x1280.jpg 1280w" `
 
turning into 

"srcset": {
   "360w": "img-240x360.jpg", 
   "540w": img-360x540.jpg",
   "980w": img-653x980.jpg", 
   "1280w": img-853x1280.jpg"
}


}

**Comment by jgarber623** [comment]
Building on @dshanske's comment with some uses cases pulled from [MDN's Responsive Images tutorial](https://developer.mozilla.org/en-US/docs/Learn/HTML/Multimedia_and_embedding/Responsive_images). URLs are resolved based on a hypothetical source page/base URL at `http://example.com/`.

### Simple `<img>` element

```html
<img src="elva-fairy-800w.jpg" alt="Elva dressed as a fairy">
```

```json
"photo": [
  {
    "value": "http://example.com/elva-fairy-800w.jpg",
    "alt": "Elva dressed as a fairy"
  }
]
```

### `<img>` element with `srcset` attribute

```html
<img srcset="elva-fairy-480w.jpg 480w,
             elva-fairy-800w.jpg 800w"
     sizes="(max-width: 600px) 480px,
            800px"
     src="elva-fairy-800w.jpg"
     alt="Elva dressed as a fairy">
```

```json
"photo": [
  {
    "value": "http://example.com/elva-fairy-800w.jpg",
    "srcset": {
      "480w": "http://example.com/elva-fairy-480w.jpg",
      "800w": "http://example.com/elva-fairy-800w.jpg"
    },
    "alt": "Elva dressed as a fairy"
  }
]
```

### `<img>` element with `srcset` attribute and implied descriptor

```html
<img srcset="elva-fairy-320w.jpg,
             elva-fairy-480w.jpg 1.5x,
             elva-fairy-640w.jpg 2x"
     src="elva-fairy-640w.jpg"
     alt="Elva dressed as a fairy">
```

```json
"photo": [
  {
    "value": "http://example.com/elva-fairy-640w.jpg",
    "srcset": {
      "1x": "http://example.com/elva-fairy-320w.jpg",
      "1.5x": "http://example.com/elva-fairy-480w.jpg",
      "2x": "http://example.com/elva-fairy-640w.jpg"
    },
    "alt": "Elva dressed as a fairy"
  }
]
```

The above example demonstrates how to handle this note [from the documentation](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/img#Attributes):

> If no descriptor is specified, the source is assigned the default descriptor: `1x`.

**Comment by jgarber623** [comment]
Don't look now, but there's also the `imagesrcset` attribute on `<link>` elements: https://html.spec.whatwg.org/multipage/indices.html#attributes-3

Same values permitted: a comma-separated list of [image candidate strings](https://html.spec.whatwg.org/multipage/images.html#image-candidate-string).

**Comment by dshanske** [comment]
The conclusion at the end of today's Microformats online session was that someone should try implementing this behind an experimental flag to move forward with this idea.

**Comment by jgarber623** [comment]
I've just published MicroMicro v3.1.0 ([RubyGems](https://rubygems.org/gems/micromicro), [GitHub release notes](https://github.com/jgarber623/micromicro/releases/tag/v3.1.0)) which implements `srcset` parsing. 🎉 

The implementation matches the examples in [my comment above](#issuecomment-619023503) along with feedback in the `#microformats` IRC/Slack channel.

**Comment by jgarber623** [comment]
Following up on @tantek's [request for implementation details](https://chat.indieweb.org/microformats/2022-09-24/1664049153541000), here's where we collectively landed on parsing an `img` element's `srcset` attribute. This logic builds upon the microformats2 parsing spec's [parse an img element for src and alt](https://microformats.org/wiki/microformats2-parsing#parse_an_img_element_for_src_and_alt) which applies to `u-*` property parsing ([link](https://microformats.org/wiki/microformats2-parsing#parsing_a_u-_property)) and implied photo property parsing ([link](https://microformats.org/wiki/microformats2-parsing#parsing_for_implied_properties)).

For reference, relevant classes and methods from [MicroMicro](https://github.com/jgarber623/micromicro)'s implementation:

- [`UrlPropertyParser#value`](https://github.com/jgarber623/micromicro/blob/main/lib/micro_micro/parsers/url_property_parser.rb#L18-L31)
- [`ImpliedPhotoPropertyParser#value`](https://github.com/jgarber623/micromicro/blob/main/lib/micro_micro/parsers/implied_photo_property_parser.rb#L13-L28)
- [`ImageElementParser`](https://github.com/jgarber623/micromicro/blob/main/lib/micro_micro/parsers/image_element_parser.rb)

Based on the existing specification, feedback in IRC/Slack, Mozilla's [HTMLImageElement.srcset documentation](https://developer.mozilla.org/en-US/docs/Web/API/HTMLImageElement/srcset), and [the WHATWG HTML specification's `srcset` attribute documentation](https://html.spec.whatwg.org/multipage/images.html#srcset-attribute), I implemented the following (presented as an update to the existing microformats2 parsing specification):

> **parse an img element for src, srcset, and alt**
>
> To parse an `img` element for `src`, `srcset`, and `alt` attributes:
>
> - if `img[alt]` or `img[srcset]`
>   - return a new `{}` structure with
>     - `value`: the element's `src` attribute as a normalized absolute URL, following the containing document's language's rules for resolving relative URLs (e.g. in HTML, use the current URL context as determined by the page, and first `<base>` element, if any).
>     - `srcset`: if present, the element's `srcset` attribute value parsed as a new `{}` structure with key/value pairs representing the attribute value's valid image candidates. See "parse a `srcset` attribute value" below.
>     - `alt`: if present, the element's `alt` attribute value, otherwise omit this key.
> - else
>   - return the element's `src` attribute as a normalized absolute URL, following the containing document's language's rules for resolving relative URLs (e.g. in HTML, use the current URL context as determined by the page, and first `<base>` element, if any).
>
> **parse a `srcset` attribute value**
>
> - create a new `{}` structure
> - split the `srcset` attribute's value on the `,` (comma) character
> - for each image candidate string in the resulting array:
>   - strip leading and trailing whitespace
>   - parse the URL from the image candidate string as all non-whitespace characters from the beginning of the string until either a whitespace character or the end of the string (whichever is first)
>   - if remaining non-whitespace characters exist
>     - parse the condition descriptor from the image candidate string as any remaining non-whitespace characters
>   - if no remaining non-whitespace characters exist
>     - assign the default condition descriptor `1x`
>   - prepare a new key/value pair:
>     - key: return the candidate descriptor (e.g. `1x`)
>     - value: return the normalized absolute URL, following the containing document's language's rules for resolving relative URLs (e.g. in HTML, use the current URL context as determined by the page, and first <base> element, if any).
>   - if the `{}` structure **does not** contain a key equal to the candidate descriptor (e.g. `1x`), add the new key/value pair to the `{}` structure
>   - if the `{}` structure contains a key equal to the candidate descriptor, ignore the new key/value pair

☝🏻  Feedback and refinement would be welcome.

That's a long-winded spec-y way of saying…

Take a `srcset` attribute value like this:

```
/image-480.jpg 480w, /image.jpg, /image-640.jpg 640w, /image-2x.jpg 2x, /image-nope.jpg 640w, /image-nope.jpg
```

…and return a hash like:

```json
{
  "480w": "https://jgarber.example/image-480.jpg",
  "1x": "https://jgarber.example/image.jpg",
  "640w": "https://jgarber.example/image-640.jpg",
  "2x": "https://jgarber.example/image-2x.jpg"
}
```

For a real-world example, see [this photo on my website](https://sixtwothree.org/photos/256) and the results parsed by [micromicro.cc](https://micromicro.cc/search?url=https%3A%2F%2Fsixtwothree.org%2Fphotos%2F256) (which uses v3.1.0 of MicroMicro).

The relevant snippet from the microformats2 JSON:

```json
"photo": [
  {
    "value": "https://assets.sixtwothree.org/uploads/photos/256/5394F7C7-97B5-465D-AF6D-B9CC85F98212_medium.jpg",
    "srcset": {
      "500w": "https://assets.sixtwothree.org/uploads/photos/256/5394F7C7-97B5-465D-AF6D-B9CC85F98212_small.jpg",
      "750w": "https://assets.sixtwothree.org/uploads/photos/256/5394F7C7-97B5-465D-AF6D-B9CC85F98212_medium.jpg",
      "1000w": "https://assets.sixtwothree.org/uploads/photos/256/5394F7C7-97B5-465D-AF6D-B9CC85F98212_large.jpg"
    },
    "alt": ""
  }
]
```

**Comment by jgarber623** [comment]
Also, the MicroMicro test suite includes several test files (HTML and JSON) consistent with [the microformats/tests repo](https://github.com/microformats/tests). Those files are available here: 

https://github.com/jgarber623/micromicro/tree/main/spec/support/fixtures/micromicro_test_suite

**Comment by gdog2u** [comment]
I'm curious how/if incorrect usages of `srcset` should be handled. I received some dubious advice some time ago that an appropriate way to serve webp images was to set the `src` attribute to an initial value of a more common file type, and that you can then use `srcset` to serve the webp file. This then allows the browser to serve the webp image if it can, or otherwise fallback on the initial value. While this does indeed work, the result is that I have two images in the `srcset` with no sizes defined.

An example of how many of the images on my site are displayed.
```html
<img srcset="logo.webp, logo.jpeg" src="logo.jpeg" alt="The logo for my site" class="u-photo">
```

If I understand the specification documented above by @jgarber623 correctly, then I presume that the following structure would be attempted, 
```json
"photo": [
  {
    "value": "https://my-website.com/logo.jpg",
    "srcset": {
      "1x": "https://my-website.com/logo.webp",
      "1x": "https://my-website.com/logo.jpg"
    },
    "alt": "The logo for my site"
  }
]
```

I imagine the actual result would be that only the last entry in the `srcset` would be listed at the `1x` key.

While I understand that this is explicitly against the [WHATWG specs](https://html.spec.whatwg.org/multipage/images.html#srcset-attribute:~:text=There%20must%20not%20be%20an%20image%20candidate%20string%20for%20an%20element%20that%20has%20the%20same%20width%20descriptor%20value%20as%20another%20image%20candidate%20string%27s%20width%20descriptor%20value%20for%20the%20same%20element), it is accepted behavior in both Chrome 120.0.6099.201, and Firefox 121.0. Similarly, if I got this info somewhere, surely other people did too.

Is it worth it to try and consider this possibility in these specs?
